File: cascade-screening-letters/node/src/cascadeService.ts
/**
 * Orchestrator: selector + resolver + one template run + writer per relative.
 *
 * Knows the sequence of calls but no HTTP and no filesystem. The template
 * is executed once per pedigree: the server endpoint doesn't accept a
 * per-individual target, so running it N times would burn quota for nothing.
 */

import type { EvageneApi } from './evageneClient.js';
import { composeLetter, type LetterSink } from './letterWriter.js';
import { type LetterTarget, selectAtRiskRelatives } from './relativeSelector.js';
import { resolveTemplateId } from './templateResolver.js';

export class NoAtRiskRelativesError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'NoAtRiskRelativesError';
  }
}

export interface CascadeRequest {
  readonly pedigreeId: string;
  readonly templateOverride?: string;
  readonly dryRun: boolean;
}

export interface CascadeResult {
  readonly targets: readonly LetterTarget[];
  readonly writtenPaths: readonly string[];
}

export interface CascadeServiceOptions {
  readonly client: EvageneApi;
  readonly sink: LetterSink;
}

export class CascadeService {
  private readonly client: EvageneApi;
  private readonly sink: LetterSink;

  constructor(options: CascadeServiceOptions) {
    this.client = options.client;
    this.sink = options.sink;
  }

  async generateLetters(request: CascadeRequest): Promise<CascadeResult> {
    const targets = await this.selectTargets(request.pedigreeId);
    if (request.dryRun) {
      return { targets, writtenPaths: [] };
    }
    const templateBody = await this.renderTemplate(request);
    const writtenPaths = targets.map((target, index) =>
      this.sink.write(composeLetter(target, templateBody, index + 1)),
    );
    return { targets, writtenPaths };
  }

  private async selectTargets(pedigreeId: string): Promise<readonly LetterTarget[]> {
    const register = await this.client.fetchRegister(pedigreeId);
    if (register.probandId === null) {
      throw new NoAtRiskRelativesError(
        `Pedigree ${pedigreeId} has no designated proband; ` +
          `set one in the Evagene web app before running this tool.`,
      );
    }
    const targets = selectAtRiskRelatives(register);
    if (targets.length === 0) {
      throw new NoAtRiskRelativesError(
        `Pedigree ${pedigreeId} has no first- or second-degree relatives with a display name recorded.`,
      );
    }
    return targets;
  }

  private async renderTemplate(request: CascadeRequest): Promise<string> {
    const templateId = await resolveTemplateId(this.client, request.templateOverride);
    return this.client.runTemplate(templateId, request.pedigreeId);
  }
}

File: cascade-screening-letters/python/src/cascade_letters/cascade_service.py
"""Orchestrator: selector + resolver + one template run + writer per relative.

Knows the sequence of calls but no HTTP and no filesystem.  The template
is executed once per pedigree — the endpoint does not accept a
per-individual target, so running it N times would produce N copies of
the same text and burn quota for nothing.
"""

from __future__ import annotations

from dataclasses import dataclass

from .evagene_client import EvageneApi
from .letter_writer import LetterSink, compose_letter
from .relative_selector import LetterTarget, select_at_risk_relatives
from .template_resolver import resolve_template_id


class NoAtRiskRelativesError(RuntimeError):
    """Raised when the pedigree has no first- or second-degree relatives to write to."""


@dataclass(frozen=True)
class CascadeRequest:
    pedigree_id: str
    template_override: str | None
    dry_run: bool


@dataclass(frozen=True)
class CascadeResult:
    targets: tuple[LetterTarget, ...]
    written_paths: tuple[str, ...]


class CascadeService:
    def __init__(self, *, client: EvageneApi, sink: LetterSink) -> None:
        self._client = client
        self._sink = sink

    def generate_letters(self, request: CascadeRequest) -> CascadeResult:
        targets = self._select_targets(request.pedigree_id)
        if request.dry_run:
            return CascadeResult(targets=tuple(targets), written_paths=())

        template_body = self._render_template(request)
        written = tuple(
            self._sink.write(compose_letter(target, template_body, index))
            for index, target in enumerate(targets, start=1)
        )
        return CascadeResult(targets=tuple(targets), written_paths=written)

    def _select_targets(self, pedigree_id: str) -> list[LetterTarget]:
        register = self._client.fetch_register(pedigree_id)
        if register.proband_id is None:
            raise NoAtRiskRelativesError(
                f"Pedigree {pedigree_id} has no designated proband; "
                "set one in the Evagene web app before running this tool."
            )
        targets = select_at_risk_relatives(register)
        if not targets:
            raise NoAtRiskRelativesError(
                f"Pedigree {pedigree_id} has no first- or second-degree relatives "
                "with a display name recorded."
            )
        return targets

    def _render_template(self, request: CascadeRequest) -> str:
        template_id = resolve_template_id(self._client, request.template_override)
        return self._client.run_template(template_id, request.pedigree_id)
